lib/charms/vault_k8s/v0/vault_managers.py

```diff
@@ -0,0 +1,372 @@
+"""Library for managing Vault Charm features.
+
+This library encapsulates the business logic for managing the Vault service and
+its associated integrations within the context of our charms.
+"""
+
+# The unique Charmhub library identifier, never change it
+import logging
+from dataclasses import dataclass
+
+from charms.vault_k8s.v0.juju_facade import JujuFacade
+from charms.vault_k8s.v0.vault_autounseal import (
+    AutounsealDetails,
+    VaultAutounsealProvides,
+    VaultAutounsealRequires,
+)
+from charms.vault_k8s.v0.vault_client import (
+    AppRole,
+    SecretsBackend,
+    Token,
+    VaultClient,
+    VaultClientError,
+)
+from charms.vault_k8s.v0.vault_tls import File, VaultTLSManager
+from ops import CharmBase, Model, Relation, SecretNotFoundError
+
+LIBID = "4a8652e06ecb4eb28c5fdbf220d126bb"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+
+AUTOUNSEAL_POLICY = """path "{mount}/encrypt/{key_name}" {{
+    capabilities = ["update"]
+}}
+
+path "{mount}/decrypt/{key_name}" {{
+    capabilities = ["update"]
+}}
+"""
+
+
+class LogAdapter(logging.LoggerAdapter):
+    """Adapter for the logger to prepend a prefix to all log lines."""
+
+    prefix = "vault_managers"
+
+    def process(self, msg, kwargs):
+        """Decides the format for the prepended text."""
+        return f"[{self.prefix}] {msg}", kwargs
+
+
+logger = LogAdapter(logging.getLogger(__name__), {})
+
+
+@dataclass
+class VaultAutounsealRelationDetails:
+    """Computes the auto-unseal related values for a relation.
+
+    This class is used to compute the static details for a vault-autounseal
+    relation, such as the key name, policy name, and approle name. These values
+    are all based on the relation ID.
+
+    This class provides a central place to manage the naming conventions for
+    the auto-unseal functionality.
+    """
+
+    relation: Relation
+
+    @property
+    def key_name(self) -> str:
+        """Return the key name for the relation."""
+        return str(self.relation.id)
+
+    @property
+    def policy_name(self) -> str:
+        """Return the policy name for the relation."""
+        return f"charm-autounseal-{self.relation.id}"
```

Comment:
This prefix shows up a lot in this file, should we use a variable?

Reply:
There's a couple different cases, but you're right each one is used in a couple of other places.

I'm not sure this class serves its original purpose anymore, so I will revisit it.

Reply:
I took a stab at this here: https://github.com/canonical/vault-k8s-operator/pull/534/commits/4adb4c949ae9806da229804583c7eb83efdbab0d

Let me know your thoughts. I don't love the name.